### PR TITLE
[posix] resolve the RCP timeout issue

### PR DESCRIPTION
**Pull request by zhanglongxia**
When calling `scan energy` command continuously, the SPI based RCP Host
crashes and the log outputs RCP timeout error message. The root cause is
that the interval between two Spinel commands is very small. After
processing the previous SPI packet, the RCP needs to take a certain amount
of time to re-set the TX/RX buffer of the SPI slave controller. During
the process of RCP setting the SPI slave TX/RX buffers, the Host starts
an SPI transfer, and then the RCP side loses the packet at this time.

This PR checks the result of the SPI transmission when sending the Spinel
command. If the SPI transmission is failed, the Host retransmits the SPI
frame. This PR also change the fd of the interrupt pin to O_NONBLOCK mode.

### CODE REVIEW

**Review by bukepo — APPROVED** [review]

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

with nits

File: src/posix/platform/spi_interface.cpp [hunk 1]
```diff
@@ -830,7 +835,12 @@ otError SpiInterface::SendFrame(const uint8_t *aFrame, uint16_t aLength)
     mSpiTxIsReady     = true;
     mSpiTxPayloadSize = aLength;
 
-    IgnoreError(PushPullSpi());
+    while (count < kMaxNumRetries)
+    {
+        VerifyOrExit((error = PushPullSpi()) != OT_ERROR_NONE);
+        usleep(1000);
+        count++;
+    }
```
Comment (abtink) [inline src/posix/platform/spi_interface.cpp:843]: I am not sure of the intention behind this change:

- We are basically calling `PushPull()` 5 times in row each time adding 1msec delay.
- This is then adding 5msec delay in main run loop of OT stack for every SPI transaction which is not good.
- Once we call `PushPull()`, if we have any pending tx frame we sent to peripheral and if it indicates that it accepted it, we clear tx frame. 
- Then the follow up 4 calls of `PushPull()` will simply send empty transactions.
- If the peripheral could not accept the frame, then we would anyway retry the same tx frame in next time we get to `Process()` and do `PushPull()`.

Comment (simonlingoogle) [reply src/posix/platform/spi_interface.cpp:843]: I think we quit immediately if PushPullSpi is successfully. So, in most cases, we don't retry at all, hence no delay.

**Comment by zhanglongxia** [comment]
> Thansk @zhanglongxia.
> 
> I have some questions about this PR. (I am using the terms peripheral/central for SPI slave/master below instead)
> 
> > The root cause is that the interval between two Spinel commands is very small.
> > After processing the previous SPI packet, the RCP needs to take a certain amount
> > of time to re-set the TX/RX buffer of the SPI slave controller. During
> > the process of RCP setting the SPI slave TX/RX buffers, the Host starts
> > an SPI transfer, and then the RCP side loses the packet at this time.
> 
> If this is the behavior/cause, then I think there is a bug/issue on the RCP SPI side (which we should then address). The SPI peripheral design and implementation should not require delay in between transitions to function.
> 
> * In the SPI protocol in a a transition either side indicates how much data it could accept, so the other side can tell whethere what it sent was received or it need to send it again.
> * So if the peripheral is not yet ready to receive a next fram from central, it should indicate this in the transaction.
> * Once a transaction is finished, the SPI driver should reset the rx/tx buffers to empty which is equivalent to it indicating it won't accept a frame in next transaction. These are described in the SPI platform APIs as well.

This issue is not caused by the SPI protocol itself. I think it is caused by the SPI slave controller hardware design.
We are using the nrf52811 chip as the RCP chip. The datasheet of the nrf52811 says
```
The SPI slave will try to acquire the semaphore when CSN goes low. 
If the SPI slave does not manage to acquire the semaphore at this
point, the transaction will be ignored. This means that all incoming
data on MOSI will be discarded, and the DEF (default) character will
be clocked out on the MISO line throughout the whole transaction. 
```
Based on this description, the peripheral of the nrf52811 has a chance to drop the SPI frame and output the DEF characters to the central. In my experiments, the SPI controller did send DEF characters when the host calls the `PushPullSpi()`.  In this case, the RCP lost the SPI frame, and it won't send the response to the host.
On the RCP side, it is hard to fix this issue through the software. On the Host side, what we can do is re-transmit the SPI frame.

**Comment by abtink** [comment]
> This issue is not caused by the SPI protocol itself. I think it is caused by the SPI slave controller hardware design. We are using the nrf52811 chip as the RCP chip. The datasheet of the nrf52811 says
> 
> ```
> The SPI slave will try to acquire the semaphore when CSN goes low. 
> If the SPI slave does not manage to acquire the semaphore at this
> point, the transaction will be ignored. This means that all incoming
> data on MOSI will be discarded, and the DEF (default) character will
> be clocked out on the MISO line throughout the whole transaction. 
> ```
> 
> Based on this description, the peripheral of the nrf52811 has a chance to drop the SPI frame and output the DEF characters to the central. In my experiments, the SPI controller did send DEF characters when the host calls the `PushPullSpi()`. In this case, the RCP lost the SPI frame, and it won't send the response to the host. On the RCP side, it is hard to fix this issue through the software. On the Host side, what we can do is re-transmit the SPI frame.


I think that fact that SPI peripheral may not be able to receive at all times is normal and expected.  It is not a bug on RCP SPI side. The description from datasheet is also very reasonable and what I would expect.

Basically if the peripheral is not ready to receive (for whatever reason), it should ignore the bytes that central is sending and send a constant byte (`0xff` or `0x00`) back to the central (typical hw implemnation is to keep the MISO pin high which maps to sending `0xff` byte). 

The Spinel SPI protocol should detect and handle this, basically seeing that peripheral send back all `0xff` causes the whole transaction to be considered invalid and retried on central.

**Comment by zhanglongxia** [comment]
The [PR7431](https://github.com/openthread/openthread/pull/7431) resolved this issue.

**Pull request closed without merging**